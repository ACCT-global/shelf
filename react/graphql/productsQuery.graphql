query Products(
    $category: String
    $specificationFilters: [String]
    $orderBy: String
    $from: Int
    $to: Int
  ) {
    products(
      category: $category
      specificationFilters: $specificationFilters
      orderBy: $orderBy
      from: $from
      to: $to
    ) {
      productId
      productName
      linkText
      description
      items {
        itemId
        name
        nameComplete
        complementName
        ean
        referenceId {
          Key
          Value
        }
        measurementUnit
        unitMultiplier
        images {
          imageId
          imageLabel
          imageTag
          imageUrl
          imageText
        }
        sellers {
          sellerId
          sellerName
          commertialOffer {
            Price
            ListPrice
            PriceWithoutDiscount
            RewardValue
            PriceValidUntil
            AvailableQuantity
          }
        }
      }
    }
}