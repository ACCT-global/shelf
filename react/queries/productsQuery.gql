query Products(
  $category: String
  $collection: String
  $specificationFilters: [String]
  $orderBy: String
  $from: Int
  $to: Int
  $hideUnavailableItems: Boolean = false
) {
  products(
    category: $category
    collection: $collection
    specificationFilters: $specificationFilters
    orderBy: $orderBy
    from: $from
    to: $to
    hideUnavailableItems: $hideUnavailableItems
  ) {
    cacheId
    productId
    productName
    productReference
    description
    categories
    # categoryTree {
    #   name
    #   href
    # }
    link
    linkText
    brand
    brandId
    items {
      name
      itemId
      referenceId {
        Value
      }
      images {
        imageUrl
        imageTag
      }
      sellers {
        sellerId
        commertialOffer {
          Installments {
            Value
            InterestRate
            TotalValuePlusInterestRate
            NumberOfInstallments
            Name
          }
          AvailableQuantity
          Price
          ListPrice
        }
      }
    }
    productClusters {
      id
      name
    }
  }
}
