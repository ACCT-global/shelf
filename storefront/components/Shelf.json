{
  "resourceBindings": [
    {
      "locator": "search@vtex.storefront-sdk",
      "relativePath": "/{{ account }}/products",
      "queryParams": {
        "category": "{{ settings.category }}",
        "collection": "{{ settings.collection }}",
        "pageSize": "{{ settings.quantity }}"
      },
      "bindTo": "products@vtex"
    }
  ],
  "assets": [
    "Shelf.js",
    "editors/index.js"
  ]
}
