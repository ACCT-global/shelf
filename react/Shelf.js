import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { graphql } from 'react-apollo'

import Slider from 'react-slick'
import ShelfItem from './ShelfItem'
import Arrow from './Arrow'
import Dots from './Dots'
import Spinner from '@vtex/styleguide/lib/Spinner'
import Card from '@vtex/styleguide/lib/Card'
import spinnerStyle from '@vtex/styleguide/lib/Spinner/style.css'

import productsQuery from './graphql/productsQuery.gql'

/**
 * Shelf Component. Shows a collection of products.
 */
class Shelf extends Component {
  configureSettings() {
    const { arrows, scroll } = this.props
    
    return {
      slidesToShow: 5,
      slidesToScroll: scroll === 'BY_PAGE' ? 5 : 1,
      dots: true,
      arrows,
      nextArrow: <Arrow color="#000" />,
      prevArrow: <Arrow color="#000" />,
      infinite: false,
      appendDots: dots => <Dots color="#000" dots={dots} />,
      responsive: [
        {
          breakpoint: 1024,
          settings: {
            slidesToShow: 3,
            slidesToScroll: scroll === 'BY_PAGE' ? 3 : 1,
          },
        },
        {
          breakpoint: 600,
          settings: {
            slidesToShow: 1,
            slidesToScroll: 1,
          },
        },
      ],
    }
  }

  render() {
    const { data, maxItems, titleText } = this.props
    const products = !data || data['error'] ? [] : data.products
    const slideSettings = this.configureSettings()

    return (
      <div className="ml7 mr7 pv4 vtex-shelf">
        <div className="w-100 flex justify-center">
          <h1>{titleText}</h1>
        </div>
        {
          data.loading && (
            <div className="w-100 flex justify-center">
              <div className="w3 ma0">
                <Spinner style={spinnerStyle} />
              </div>
            </div>
          )}
        {
          !data.loading && products && (
            <Slider {...slideSettings}>
              {products.slice(0, maxItems).map(item => {
                return (
                  <div key={item.productId} className="pa4">
                    <Card>
                      <ShelfItem {...item} imageHeight={200} />
                    </Card>
                  </div>
                )
              })}
            </Slider>
          )}
      </div>
    )
  }
}

Shelf.schema = {
  title: 'Shelf',
  description: 'A product shelf featuring a collection',
  type: 'object',
  properties: {
    category: {
      title: 'Category',
      type: 'number'
    },
    collection: {
      title: 'Collection',
      type: 'number',
    },
    orderBy: {
      title: 'List Ordenation',
      type: 'string',
      enum: ['OrderByTopSaleDESC', 'OrderByPriceDESC', 'OrderByPriceASC'],
      enumNames: ['Sales', 'Price, descending', 'Price, ascending'],
      default: 'OrderByTopSaleDESC',
    },
    maxItems: {
      title: 'Max Items',
      type: 'number',
      default: 10,
    },
    scroll: {
      title: 'Scroll Type',
      type: 'string',
      enum: ['BY_PAGE', 'ONE_BY_ONE'],
      enumNames: ['By Page', 'One By One'],
      default: 'BY_PAGE',
    },
    arrows: {
      title: 'Arrows',
      type: 'boolean',
      default: true,
    },
    titleText: {
      title: 'Title Text',
      type: 'string',
      default: 'Default Title',
    },
  },
}

Shelf.defaultProps = {
  maxItems: 10,
  scroll: 'BY_PAGE',
  arrows: true
}

Shelf.propTypes = {
  /** The graphql data response. */
  data: PropTypes.shape({
    products: PropTypes.arrayOf(PropTypes.shape({
      productId: PropTypes.string.isRequired,
      productName: PropTypes.string.isRequired,
      items: PropTypes.arrayOf(PropTypes.shape({
        images: PropTypes.arrayOf(PropTypes.shape({
          imageUrl: PropTypes.string.isRequired
        })).isRequired,
        sellers: PropTypes.arrayOf(PropTypes.shape({
          commertialOffer: PropTypes.shape({
            Price: PropTypes.number.isRequired
          })
        })).isRequired
      })),
    })),
  }).isRequired,
  /** The Category Id. */
  category: PropTypes.number,
  /** The Collection Id. */
  collection: PropTypes.number,
  /** The Ordenation Type. */
  orderBy: PropTypes.string,
  /** Maximum number of items in the shelf. */
  maxItems: PropTypes.number.isRequired,
  /** The scroll options. */
  scroll: PropTypes.string.isRequired,
  /** The Collection Id. */
  arrows: PropTypes.bool.isRequired,
  /** The text value of the title. */
  titleText: PropTypes.string,
}

const options = {
  options: ({
    category,
    collection,
    orderBy,
    maxItems = Shelf.defaultProps.maxItems,
  }) => ({
    variables: {
      category,
      collection,
      orderBy,
      from: 0,
      to: maxItems - 1,
      specificationFilters: [],
    },
    ssr: false,
  }),
}

export default graphql(productsQuery, options)(Shelf)
