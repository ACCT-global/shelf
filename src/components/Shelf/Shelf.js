import { editable } from 'vtex-editor';
import React from 'react';
import Immutable from 'immutable';
import './Shelf.less';
import ShelfPlaceholder from './ShelfPlaceholder';
import ShelfSlider from './ShelfSlider';

@editable({
  name: 'Shelf@vtex.shelf',
  title: 'Shelf'
})
class Shelf extends React.Component {
  static defaultProps = {
    settings: Immutable.fromJS({
      title: 'Destaques',
      category: 'fera-fashion',
      quantity: 3
    })
  }

  render() {
    if (!this.props.settings) {
      return (
        <ShelfPlaceholder title="Destaques"/>
      );
    }
    return <ShelfSlider {...this.props}/>;
  }
}

export default Shelf;
