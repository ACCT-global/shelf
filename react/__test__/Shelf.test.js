/* eslint-env jest */
import React from 'react'
import { render } from 'react-testing-library'
import { MockedProvider } from 'react-apollo/test-utils'

import Shelf from '../Shelf'
import productsQuery from '../graphql/productsQuery.gql'

describe('Shelf component', () => {
  let wrapper

  beforeEach((done) => {
    const mockedProducts = [
      {
        productId: '1',
        productName: 'Product1',
        brand: 'brand',
        link: 'http://mylink.com',
        items: [{
          name: 'name',
          referenceId: [{
            Value: 'ref1234',
          }],
          images: [{
            imageUrl: '//www.allfree-clipart.com/Business/computer.jpg',
            imageTag: '<img src="//www.allfree-clipart.com/Business/computer.jpg">',
          }],
          sellers: [{
            commertialOffer: {
              Price: 400,
              ListPrice: 200,
            },
          }],
        }],
      },
      {
        productId: '2',
        productName: 'Product2',
        brand: 'brand',
        link: 'http://mylink.com',
        items: [{
          name: 'name',
          referenceId: [{
            Value: 'ref1234',
          }],
          images: [{
            imageUrl: '//www.allfree-clipart.com/Business/computer.jpg',
            imageTag: '<img src="//www.allfree-clipart.com/Business/computer.jpg"',
          }],
          sellers: [{
            commertialOffer: {
              Price: 400,
              ListPrice: 200,
            },
          }],
        }],
      },
      {
        productId: '3',
        productName: 'Product3',
        brand: 'brand',
        link: 'http://mylink.com',
        items: [{
          name: 'name',
          referenceId: [{
            Value: 'ref1234',
          }],
          images: [{
            imageUrl: '//www.allfree-clipart.com/Business/computer.jpg',
            imageTag: '<img src="//www.allfree-clipart.com/Business/computer.jpg"',
          }],
          sellers: [{
            commertialOffer: {
              Price: 400,
              ListPrice: 200,
            },
          }],
        }],
      },
    ]

    wrapper = render(
      <MockedProvider
        mocks={[
          {
            request: { query: productsQuery, variables: {
              specificationFilters: [],
              from: 0,
              to: 9,
            } },
            result: { data: { products: mockedProducts } },
          },
        ]}
      >
        <Shelf />
      </MockedProvider>
    )

    // necessary because we need to wait for the graphql
    // response, even if it's a mocked one
    setTimeout(() => {
      done()
    }, 0)
  })

  it('should be rendered', () => {
    expect(wrapper).toBeDefined()
  })

  it('should match snapshot', () => {
    expect(wrapper.container).toMatchSnapshot()
  })

  it('should render 3 slide items', () => {
    expect(
      wrapper.container.querySelectorAll('.vtex-shelf').length
    ).toBe(1)
    expect(
      wrapper.container.querySelectorAll('.slick-slider').length
    ).toBe(1)
    expect(
      wrapper.container.querySelectorAll('.slick-slide').length
    ).toBe(6) // 6 because the slide is infinite
  })
})
