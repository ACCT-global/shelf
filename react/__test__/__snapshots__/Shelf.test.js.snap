// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shelf component should match snapshot 1`] = `
<div>
  <div
    class="ml7 mr7 pv4 vtex-shelf"
  >
    <div
      class="vtex-shelf__title-content w-100 flex justify-center"
    >
      <h1
        class="vtex-shelf__title-text"
      >
         
      </h1>
    </div>
    <div
      class="regular slider slick-initialized slick-slider"
    >
      
      <div
        class="slick-list"
      >
        <div
          class="slick-track"
          style="opacity: 1;"
        />
      </div>
      
      
    </div>
  </div>
</div>
`;
